File: ios/App/App/Call/NativePresenceLeaseRenewOwner.swift
import Foundation

/**
 * Session-bound Native presence lease renew (D Hybrid / iOS).
 *
 * WebView-independent. NOT a 10s heartbeat clone.
 * Cadence mirrors lib/call/call-active-presence.ts (SHADOW_LEASE_TTL / 2).
 *
 * Sparse renew is best-effort while CallKit/runtime session is connected.
 * Do NOT claim guaranteed background timer cadence under iOS suspension.
 *
 * Production end authority remains legacy_hb (LEASE CUTOVER: NO).
 */
enum NativePresenceLeaseRenewOwner {
  /// Mirror CALL_PRESENCE_SHADOW_LEASE_TTL_MS
  static let shadowLeaseTtlMs: Int = 300_000
  /// Mirror CALL_PRESENCE_NATIVE_RENEW_INTERVAL_MS
  static let renewIntervalMs: Int = shadowLeaseTtlMs / 2

  typealias RenewCallback = (_ ok: Bool, _ status: Int, _ error: String?) -> Void
  typealias RenewTransport = (_ callId: String, _ completion: @escaping RenewCallback) -> Void
  typealias LiveGate = (_ callId: String) -> Bool

  private static let lock = NSLock()
  private static var activeCallId: String?
  private static var activeGeneration: Int = 0
  private static var generationCounter: Int = 0
  private static var sparseWorkItem: DispatchWorkItem?
  private static var transport: RenewTransport?
  private static var liveGate: LiveGate?
  private static var renewCountForTests: Int = 0
  private static let queue = DispatchQueue(label: "com.dibay.native.presence.renew")

  static func start(callId: String, renewTransport: @escaping RenewTransport, gate: @escaping LiveGate) {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !sid.isEmpty else { return }
    let gen: Int
    lock.lock()
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    gen = activeGeneration
    activeCallId = sid
    transport = renewTransport
    liveGate = gate
    renewCountForTests = 0
    lock.unlock()
    DibayCallLog.info(
      "presence_renew_start",
      sessionId: sid,
      detail: "intervalMs=\(renewIntervalMs)"
    )
    fireRenew(sid: sid, gen: gen, source: "immediate_connected")
  }

  static func stop(callId: String, reason: String) {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    lock.lock()
    defer { lock.unlock() }
    guard !sid.isEmpty, activeCallId == sid else { return }
    DibayCallLog.info("presence_renew_stop", sessionId: sid, detail: "reason=\(reason)")
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    activeCallId = nil
    transport = nil
    liveGate = nil
  }

  static func isActiveForTests(_ callId: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return activeCallId == callId.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  static func renewCountForTestsValue() -> Int {
    lock.lock()
    defer { lock.unlock() }
    return renewCountForTests
  }

  static func resetForTests() {
    lock.lock()
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    activeCallId = nil
    transport = nil
    liveGate = nil
    renewCountForTests = 0
    lock.unlock()
  }

  /// Test helper — run pending sparse renew immediately (does not prove BG cadence).
  static func flushSparseForTests() {
    lock.lock()
    let item = sparseWorkItem
    sparseWorkItem = nil
    lock.unlock()
    item?.cancel()
    item?.perform()
  }

  private static func fireRenew(sid: String, gen: Int, source: String) {
    let t: RenewTransport?
    let gate: LiveGate?
    lock.lock()
    guard activeCallId == sid, gen == activeGeneration else {
      lock.unlock()
      return
    }
    t = transport
    gate = liveGate
    lock.unlock()
    guard let t, let gate else { return }
    guard gate(sid) else {
      stop(callId: sid, reason: "not_live_gate")
      return
    }
    t(sid) { ok, status, error in
      lock.lock()
      guard activeCallId == sid, gen == activeGeneration else {
        lock.unlock()
        return
      }
      renewCountForTests += 1
      lock.unlock()
      if !ok {
        let err = error ?? ""
        if err == "not_live" || err.contains("not_live") {
          stop(callId: sid, reason: "server_not_live")
          return
        }
        DibayCallLog.info(
          "presence_renew_failed",
          sessionId: sid,
          detail: "err=\(err) source=\(source) status=\(status)"
        )
      } else {
        DibayCallLog.info("presence_renew_ok", sessionId: sid, detail: "source=\(source)")
      }
      scheduleSparse(sid: sid, gen: gen)
    }
  }

  private static func scheduleSparse(sid: String, gen: Int) {
    lock.lock()
    guard activeCallId == sid, gen == activeGeneration else {
      lock.unlock()
      return
    }
    clearPendingLocked()
    let work = DispatchWorkItem {
      fireRenew(sid: sid, gen: gen, source: "sparse_best_effort")
    }
    sparseWorkItem = work
    lock.unlock()
    // Best-effort only — do not claim guaranteed BG execution under CallKit suspension.
    queue.asyncAfter(deadline: .now() + .milliseconds(renewIntervalMs), execute: work)
  }

  private static func clearPendingLocked() {
    sparseWorkItem?.cancel()
    sparseWorkItem = nil
  }
}

File: android/app/src/main/java/com/dibay/app/call/NativePresenceLeaseRenewOwner.java
package com.dibay.app.call;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Session-bound Native presence lease renew (D Hybrid).
 *
 * WebView-independent. NOT a 10s heartbeat clone.
 * Cadence MUST mirror lib/call/call-active-presence.ts:
 *   CALL_PRESENCE_SHADOW_LEASE_TTL_MS / 2
 *
 * Production end authority remains legacy_hb (LEASE CUTOVER: NO).
 */
public final class NativePresenceLeaseRenewOwner {
  private static final String TAG = "DibayNativePresenceRenew";

  /** Mirror CALL_PRESENCE_SHADOW_LEASE_TTL_MS */
  public static final long SHADOW_LEASE_TTL_MS = 300_000L;

  /** Mirror CALL_PRESENCE_NATIVE_RENEW_INTERVAL_MS (= TTL / 2) */
  public static final long RENEW_INTERVAL_MS = SHADOW_LEASE_TTL_MS / 2;

  public interface RenewCallback {
    void onDone(boolean ok, int status, String error);
  }

  public interface RenewTransport {
    void renew(Context app, String callId, RenewCallback callback);
  }

  public interface LiveGate {
    /** True only while this session is still Native CONNECTED / renew-eligible. */
    boolean isRenewLive(String callId);
  }

  private static final Object LOCK = new Object();
  private static final Handler MAIN = new Handler(Looper.getMainLooper());
  private static final AtomicInteger GENERATION = new AtomicInteger(0);

  private static String activeCallId = null;
  private static int activeGeneration = 0;
  private static Runnable pendingSparse = null;
  private static RenewTransport transport = null;
  private static LiveGate liveGate = null;
  private static Context appContext = null;
  private static int renewCountForTests = 0;

  private NativePresenceLeaseRenewOwner() {}

  public static void start(
      Context context, String callId, RenewTransport renewTransport, LiveGate gate) {
    if (context == null || callId == null || callId.trim().isEmpty()) return;
    if (renewTransport == null || gate == null) return;
    Context app = context.getApplicationContext();
    String sid = callId.trim();
    int gen;
    synchronized (LOCK) {
      clearPendingLocked("replace_before_start");
      activeCallId = sid;
      activeGeneration = GENERATION.incrementAndGet();
      gen = activeGeneration;
      transport = renewTransport;
      liveGate = gate;
      appContext = app;
      renewCountForTests = 0;
    }
    Log.i(TAG, "presence_renew_start callId=" + sid + " intervalMs=" + RENEW_INTERVAL_MS);
    fireRenew(app, sid, gen, "immediate_connected");
  }

  public static void stop(String callId, String reason) {
    String sid = callId != null ? callId.trim() : "";
    synchronized (LOCK) {
      if (sid.isEmpty() || activeCallId == null || !sid.equals(activeCallId)) {
        return;
      }
      Log.i(TAG, "presence_renew_stop callId=" + sid + " reason=" + safe(reason));
      clearPendingLocked(reason);
      activeCallId = null;
      activeGeneration = GENERATION.incrementAndGet();
      transport = null;
      liveGate = null;
      appContext = null;
    }
  }

  public static boolean isActiveForTests(String callId) {
    synchronized (LOCK) {
      return callId != null && callId.trim().equals(activeCallId);
    }
  }

  public static int renewCountForTests() {
    synchronized (LOCK) {
      return renewCountForTests;
    }
  }

  public static void resetForTests() {
    synchronized (LOCK) {
      clearPendingLocked("reset_for_tests");
      activeCallId = null;
      activeGeneration = GENERATION.incrementAndGet();
      transport = null;
      liveGate = null;
      appContext = null;
      renewCountForTests = 0;
    }
  }

  /** Advance sparse timer for Robolectric tests. */
  public static void flushSparseForTests() {
    Runnable next;
    synchronized (LOCK) {
      next = pendingSparse;
    }
    if (next != null) {
      MAIN.removeCallbacks(next);
      next.run();
    }
  }

  private static void fireRenew(Context app, String sid, int gen, String source) {
    RenewTransport t;
    LiveGate gate;
    synchronized (LOCK) {
      if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) {
        return;
      }
      t = transport;
      gate = liveGate;
    }
    if (t == null || gate == null) return;
    if (!gate.isRenewLive(sid)) {
      stop(sid, "not_live_gate");
      return;
    }
    t.renew(
        app,
        sid,
        (ok, status, error) -> {
          synchronized (LOCK) {
            if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) {
              return;
            }
            renewCountForTests += 1;
          }
          if (!ok) {
            String err = error != null ? error : "";
            if ("not_live".equals(err) || err.contains("not_live")) {
              stop(sid, "server_not_live");
              return;
            }
            // Transient network: keep sparse schedule if still live
            Log.w(TAG, "presence_renew_failed callId=" + sid + " err=" + err + " source=" + source);
          } else {
            Log.i(TAG, "presence_renew_ok callId=" + sid + " source=" + source);
          }
          scheduleSparse(app, sid, gen);
        });
  }

  private static void scheduleSparse(Context app, String sid, int gen) {
    synchronized (LOCK) {
      if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) {
        return;
      }
      clearPendingLocked("reschedule");
      Runnable runnable =
          () -> {
            synchronized (LOCK) {
              if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) {
                return;
              }
              pendingSparse = null;
            }
            fireRenew(app, sid, gen, "sparse");
          };
      pendingSparse = runnable;
      MAIN.postDelayed(runnable, RENEW_INTERVAL_MS);
    }
  }

  private static void clearPendingLocked(String reason) {
    if (pendingSparse != null) {
      MAIN.removeCallbacks(pendingSparse);
      pendingSparse = null;
    }
  }

  private static String safe(String value) {
    return value == null ? "" : value.trim();
  }
}
